### PR TITLE
Require value types for iterables

### PR DESCRIPTION
**Pull request by thewilkybarkid**
This is a PR to require that the value types of iterables are defined (eg `array<string>` rather than just `array`).

I've implemented return types, but will hold for feedback. The only concern I can see is that it's not possible to allow defining `@return array<mixed>` (which _is_ defining a value type; can't see how to reliably distinguish between implicit and explicit use of `mixed`).

(Refs https://github.com/libero/php-coding-standard/pull/40.)

### CODE REVIEW

**Comment by ondrejmirtes** [comment]
Hi,
I think you'd be reliably able to add this feature to existing rules, e.g. MissingFunctionParameterTypehintRule, MissingFunctionReturnTypehintRule, MissingMethodParameterTypehintRule, MissingMethodReturnTypehintRule.

Just ask `Type::isIterable()->yes()` and then `Type::getIterableValueType()`.

The main hurdle is that it differentiates between implicit and explicit mixed, but the `array` is currently considered an explicit mixed. So this would be helpful only for native `array` typehint without any phpDoc, which is a good start.

**Comment by ondrejmirtes** [comment]
You could also fix the existing rules to register them to `InClassMethodNode` - the correct reflection is then available in `$scope->getFunction()` and no exceptions for traits are needed.

**Comment by CZechBoY** [comment]
btw I also tried at https://github.com/phpstan/phpstan-strict-rules/pull/35 but there were some reflection issues... is it fixed now?

**Comment by ondrejmirtes** [comment]
What reflection issues?

**Comment by CZechBoY** [comment]
https://travis-ci.org/phpstan/phpstan-strict-rules/jobs/490487289
on interface method parameter of type array, key/item is explicitly mixed.

image

**Comment by ondrejmirtes** [comment]
I said that above:

> The main hurdle is that it differentiates between implicit and explicit mixed, but the `array` is currently considered an explicit mixed. So this would be helpful only for native `array` typehint without any phpDoc, which is a good start.

So it's fine for me if the rule currently addreses native typehints without any item specification in the phpDoc.

File: src/Rules/Methods/MissingMethodReturnTypehintRule.php [hunk 1]
```diff
@@ -31,7 +34,11 @@ public function processNode(Node $node, Scope $scope): array
 			throw new \PHPStan\ShouldNotHappenException();
 		}
 
-		$methodReflection = $scope->getClassReflection()->getNativeMethod($node->name->name);
+		if (!$scope->isInTrait()) {
```
Comment (thewilkybarkid) [inline src/Rules/Methods/MissingMethodReturnTypehintRule.php]: I had to leave this in to get https://github.com/phpstan/phpstan-strict-rules/blob/a1b5b6e0e6f876ba5649c42f2092ba015f641bdb/tests/Rules/Methods/MissingMethodReturnTypehintRuleTest.php#L48-L63 to work. Would you not expect to check the trait directly?

Comment (ondrejmirtes) [reply src/Rules/Methods/MissingMethodReturnTypehintRule.php]: Please follow my advice I gave here: https://github.com/phpstan/phpstan-strict-rules/pull/70#issuecomment-461736870

Comment (thewilkybarkid) [reply src/Rules/Methods/MissingMethodReturnTypehintRule.php]: I can't get that to work... I end up with the using class name but the line number from the trait (eg `42: Method MissingMethodReturnTypehint\Foo::c1() has no return typehint specified.`).

Comment (ondrejmirtes) [reply src/Rules/Methods/MissingMethodReturnTypehintRule.php]: Can you push this change as a new commit so I can review it?

Comment (thewilkybarkid) [reply src/Rules/Methods/MissingMethodReturnTypehintRule.php]: Pushed 10eee51.

Comment (ondrejmirtes) [reply src/Rules/Methods/MissingMethodReturnTypehintRule.php]: The results look fine to me, just the declaring class is wrong there - which is something we might fix in PHPStan's core, but it's not a big deal.

**Comment by thewilkybarkid** [comment]
@ondrejmirtes Thanks; I've tweaked the tests slightly to remove the override, since it's now testing the class rather than the trait directly.

**Comment by thewilkybarkid** [comment]
@ondrejmirtes Hopefully all done now.

**Comment by CZechBoY** [comment]
@ondrejmirtes Anything needs to be done in this PR?

**Comment by ondrejmirtes** [comment]
FYI I plan to move these mixed-detecting rules from strict-rules to phpstan/phpstan as a new level 6 for the new version 0.12 :) Stay tuned!

**Comment by ondrejmirtes** [comment]
I just pushed an implementation of this in `phpstan/phpstan`: https://github.com/phpstan/phpstan/commit/f58624bb52f2d014253bb6fe17f6ad45cb2b30ae

Thank you!

**Pull request closed without merging**